### PR TITLE
D8/D9 - Configurable tag overriding

### PR DESCRIPTION
**Pull request by stevector**
https://www.drupal.org/project/pantheon_advanced_page_cache/issues/2944229

### CODE REVIEW

File: src/EventSubscriber/CacheableResponseSubscriber.php [hunk 1]
```diff
@@ -21,14 +22,42 @@ class CacheableResponseSubscriber implements EventSubscriberInterface {
...
   /**
    * Constructs a new DefaultExceptionHtmlSubscriber.
    *
    * @param \Psr\Log\LoggerInterface $logger
    *   The logger service.
+   * @param \Drupal\Core\Config\ConfigFactory $configFactory
+   *   Configuration for this module.
    */
-  public function __construct(LoggerInterface $logger) {
+  public function __construct(LoggerInterface $logger, ConfigFactory $configFactory) {
```
Comment (clayfreeman) [inline src/EventSubscriber/CacheableResponseSubscriber.php]: This is the standard way of handling new constructor arguments:

```suggestion
  public function __construct(LoggerInterface $logger, ConfigFactoryInterface $config_factory = NULL) {
    if (!$config_factory instanceof ConfigFactoryInterface) {
      @trigger_error('Not passing the config factory service as the second parameter to ' . __METHOD__ . ' is deprecated in pantheon_advanced_page_cache:1.2.0 and will throw a type error in pantheon_advanced_page_cache:2.0.0. Pass an instance of \\Drupal\\Core\\Config\\ConfigFactoryInterface. See https://www.drupal.org/node/2944229', E_USER_DEPRECATED);
      $config_factory = \Drupal::service('config.factory');
    }
```

File: src/EventSubscriber/CacheableResponseSubscriber.php [hunk 2]
```diff
@@ -21,14 +22,42 @@ class CacheableResponseSubscriber implements EventSubscriberInterface {
...
     $this->logger = $logger;
+    $this->configFactory = $configFactory;
```
Comment (clayfreeman) [inline src/EventSubscriber/CacheableResponseSubscriber.php]: This change was made in compliance with Drupal coding standards. Local variables use underscores, not camel case.

```suggestion
    $this->configFactory = $config_factory;
```

File: src/EventSubscriber/CacheableResponseSubscriber.php [hunk 3]
```diff
@@ -21,14 +22,42 @@ class CacheableResponseSubscriber implements EventSubscriberInterface {
...
+   * @var \Drupal\Core\Config\ConfigFactory
+   */
+  protected $configFactory;
+
   /**
    * Constructs a new DefaultExceptionHtmlSubscriber.
    *
    * @param \Psr\Log\LoggerInterface $logger
    *   The logger service.
+   * @param \Drupal\Core\Config\ConfigFactory $configFactory
```
Comment (clayfreeman) [inline src/EventSubscriber/CacheableResponseSubscriber.php]: You should really use the interface instead:

```suggestion
   * @param \Drupal\Core\Config\ConfigFactoryInterface $config_factory
```

File: src/EventSubscriber/CacheableResponseSubscriber.php [hunk 4]
```diff
@@ -8,6 +8,7 @@
 use Symfony\Component\EventDispatcher\EventSubscriberInterface;
 use Psr\Log\LoggerInterface;
 use Drupal\Core\Logger\RfcLogLevel;
+use Drupal\Core\Config\ConfigFactory;
```
Comment (clayfreeman) [inline src/EventSubscriber/CacheableResponseSubscriber.php]:
```suggestion
use Drupal\Core\Config\ConfigFactoryInterface;
```

File: src/EventSubscriber/CacheableResponseSubscriber.php [hunk 5]
```diff
@@ -21,14 +22,42 @@ class CacheableResponseSubscriber implements EventSubscriberInterface {
    */
   protected $logger;
 
+  /**
+   * Configuration Factory.
+   *
+   * @var \Drupal\Core\Config\ConfigFactory
```
Comment (clayfreeman) [inline src/EventSubscriber/CacheableResponseSubscriber.php]:
```suggestion
   * @var \Drupal\Core\Config\ConfigFactoryInterface
```

File: src/EventSubscriber/CacheableResponseSubscriber.php [hunk 6]
```diff
@@ -21,14 +22,42 @@ class CacheableResponseSubscriber implements EventSubscriberInterface {
...
+  public function getOverrideListTagsSetting() {
+    $config = $this->configFactory->get('pantheon_advanced_page_cache.settings');
+    // Only return FALSE if this config value is really set to false.
+    // A null value should return TRUE for backwards compatibility.
+    if ($config->get('override_list_tags') === FALSE) {
+      return FALSE;
+    }
+    else {
+      return TRUE;
+    }
```
Comment (clayfreeman) [inline src/EventSubscriber/CacheableResponseSubscriber.php]: This is simpler to read, imo.

```suggestion
    return TRUE;
```

**Review by clayfreeman — CHANGES_REQUESTED** [review]
I've attached a review covering current general best practice for adding new constructor arguments for dependency injection. I've also added feedback on code style in a couple places.

Let me know what you think! I'm excited to get this change in, especially with https://www.drupal.org/node/3107058 having been merged into Drupal core.

**Comment by clayfreeman** [comment]
~~Also, you may want to test whether your update is being applied correctly; I believe you should be using a post update hook if manipulating higher-level Drupal constructs such as configuration defaults.~~ You should be okay with what you're currently doing.

File: README.md [hunk 7]
```diff
@@ -22,6 +22,16 @@ A direct way of inspecting headers is with `curl -I`. This command will make a r
 
 `curl -IH "Pantheon-Debug:1" https://dev-cache-tags-demo.pantheonsite.io/ | grep -i Surrogate-Key-Raw`
 
+## Changing Listing Tags
+
+Prior to the 1.1 release, this module would change the cache tags used on default listings.
+This changing of was done to make cache hits more likely but resulted in [confusing cache clearing behavior](https://www.drupal.org/project/pantheon_advanced_page_cache/issues/2944229).
+Sites that installed this module prior to 1.1 should uninstall and reinstall or run this command to update their settings.
```
Comment (clayfreeman) [inline README.md]:
```suggestion
Prior to the 1.2 release, this module would change the cache tags used on default listings.
This changing of was done to make cache hits more likely but resulted in [confusing cache clearing behavior](https://www.drupal.org/project/pantheon_advanced_page_cache/issues/2944229).
Sites that installed this module prior to 1.2 should uninstall and reinstall or run this command to update their settings.
```

Comment (clayfreeman) [reply README.md]: You may wish to ignore this specific suggestion; I was going by [your comment on Drupal.org](https://www.drupal.org/project/pantheon_advanced_page_cache/issues/2944229#comment-13623674).

**Comment by stevector** [comment]
Thanks @clayfreeman !

**Pull request merged**